File: js/3-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const map = new Map;
    let longest = 0, left = 0;

    for(let right = 0; right < s.length; right ++){
        char = s.charCodeAt(right);

        if(map.has(char) && map.get(char) >= left) left = map.get(char) + 1;

        map.set(char, right);

        const windowSize = right - left + 1;
        if(windowSize > longest) longest = windowSize;
    }

    return longest;
};

File: python/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_seen = {}
        get = last_seen.get
        longest = left = 0
        length = len(s)
        for right, char in enumerate(s):
            prev = get(char, -1)
            if prev >= left:
                left = prev + 1
            elif right - left >= longest:
                longest = right - left + 1
            if length - left <= longest:
                return longest
            last_seen[char] = right
        return longest
